nexus/deployment/src/planner.rs

```diff
@@ -168,6 +162,126 @@ impl<'a> Planner<'a> {
             }
         }
 
+        self.ensure_correct_number_of_nexus_zones(
+            &sleds_ineligible_for_services,
+        )?;
+
+        Ok(())
+    }
+
+    fn ensure_correct_number_of_nexus_zones(
+        &mut self,
+        sleds_ineligible_for_services: &BTreeSet<Uuid>,
+    ) -> Result<(), Error> {
+        // Bin every sled by the number of Nexus zones it currently has while
+        // counting the total number of Nexus zones.
+        let mut num_total_nexus = 0;
+        let mut sleds_by_num_nexus: BTreeMap<usize, Vec<Uuid>> =
+            BTreeMap::new();
+        for &sled_id in self.policy.sleds.keys() {
```

Comment:
This is all totally fine.  I'm just thinking out loud: would this be a lot simpler using https://doc.rust-lang.org/std/collections/struct.BinaryHeap.html?  I was figuring you'd throw all the sleds in a heap ordered by the num_nexus count and while you're still looking for another one to deploy, pull it out of the heap, deploy the zone, and put it back in with the new value.  (That assumes a min-heap.)

I'm not asking you to rewrite this -- I was just curious if you'd considered this or if you thought it could be simpler?

Reply:
I did think about that, and thought it had a couple downsides:

* There's some small machinery to create a struct that holds both the ID and the Nexus count that sorts as we expect to store in the binary heap.
* The standard binary heap doesn't (AFAICT?) have a way to say "give me all the tied-for-lowest-count values", only "give me one of the tied-for-lowest-count values". If we want to pick from the sleds with some specific criteria (e.g., total control plane services on the sled) we're going to want the former, which is what this whole binning thing is for.

That second point is building toward a hypothetical future, and is overkill / moot if we never do the smarter choice thing. But I think it's fairly likely we'll want to do something like that?